src/Measurement_Observation/haskell_model.rst

```diff
@@ -0,0 +1,129 @@
+backends use Cmm.
+
+Cmm: an imperative language that looks like LLVM IR. It supports expressions,
+statements, and it abstracts over machine primops, machine registers, stack
+usage, and calling conventions. A pass performs register assignment and stack
+layout for the target architecture. Then assembly code (textual) is generated
+for the target architecture and an external assembler program generates machine
+code (binary) for it. Finally an external linker is used to transform the
+resulting code objects into a single executable or library.
+
+Executables are linked against a runtime system (RTS) that provides primitives
+to manage:
+- memory: allocation, garbage collection...
+- scheduling: thread scheduling, blocking queues...
+- IOs: primitives to interact with the operating system
+- dynamic code loading: loading and unloading code objects at runtime.
+
+The RTS itself comes in different flavors: e.g. using multiple OS threads to execute
+Haskell code or not.
+
+All these compilation stages and the RTS provide knobs to tweak the generated
+code and the behavior of the runtime system. In particular, some probes can be
+optionally inserted in the generated code at various stages to produce different
+profiling information at runtime.
+
+
+Execution model overview
+------------------------
+
+The heap of a Haskell program contain objects that reflect its current execution
+state: suspended computations (thunks), partially applied functions (PAP),
+values (datacon and their payload), threads (TSO) and their stack, etc.
+When the computer executes the compiled code for a top-level STG binding
+(starting from ``main``), it asks for more objects to be allocated in the heap,
+for some thunks to be reduced to another heap object, or for existing objects to
+be used as arguments for function applications.
+
+The garbage collector is responsible for freeing space in the heap. It runs when
+there is not enough space left or depending on other heuristics. GHC provides
+several knobs to configure the garbage collector strategies to use and to tweak
+their properties.
+
+The garbage collector used has an impact on profiling. For an extreme example,
+if the heap size is configured to be large enough than your program never has to
+perfom garbage collection, you'll find that you spend 0% time doing garbage
+collection and 100% executing your program ("mutator" time): all the garbage
+collection occurs at once implicitly when your program exits. On the other hand,
+if you configure the heap to be very small, most of the time can be spent doing
+garbage collection even if you haven't changed anything in your program.
+It means that **you must be very aware of the RTS options you use when profiling**.
+
+Similarly, some profiling options have an impact on the size of the heap
+objects: e.g. the heap object that represents `10 :: Int64` uses 24 bytes
+instead of 16 bytes without profiling. Because of this, you may find that your
```

Comment:
```suggestion
objects: e.g. the heap object that represents `10 :: Int64` uses 24 bytes without profiling, however with profiling this same object will use 16 bytes. Because of this, you may find that your
```